onnxruntime/core/session/inference_session.cc

```diff
@@ -1733,59 +1732,87 @@ common::Status InferenceSession::Run(IOBinding& io_binding) {
 }
 
 #ifdef ENABLE_TRAINING
-common::Status InferenceSession::RunInBackgroundAndWaitForYield(RunOptions& run_options, IOBinding& io_binding,
-                                                                std::vector<OrtValue>& user_outputs) {
-  const int64_t main_thread_event_id = 0;
-  onnxruntime::contrib::OrtEventPool::GetInstance().ResetEvent(0);
+common::Status InferenceSession::RunInBackgroundAndWaitForYield(const RunOptions& run_options, IOBinding& io_binding,
+                                                                std::vector<OrtValue>& user_outputs, int64_t& run_id) {
+  std::promise<void> setup_promise;
+  std::future<void> setup_future = setup_promise.get_future();
+
+  auto bg_thread = std::thread([&](std::future<void> setup_future) {
+    // wait until task is properly setup
+    setup_future.get();
 
-  task_.terminate_flag_ = &(run_options.terminate);
-  task_.bg_thread_promise_ = std::promise<Status>();
-  task_.bg_thread_future_ = task_.bg_thread_promise_.get_future();
-  task_.bg_thread_ = std::thread([&](std::promise<common::Status> result_promise) {
-    common::Status s = Run(run_options, io_binding.GetInputNames(), io_binding.GetInputs(), io_binding.GetOutputNames(),
-                           &io_binding.GetOutputs(), &io_binding.GetOutputsDeviceInfo());
+    common::Status status = Run(run_options, io_binding.GetInputNames(), io_binding.GetInputs(), io_binding.GetOutputNames(),
+                                &io_binding.GetOutputs(), &io_binding.GetOutputsDeviceInfo());
 
-    result_promise.set_value(s);
+    onnxruntime::contrib::OrtTasks::GetInstance().SetStatus(status);
 
     // signal main thread for background thread completion
-    const int64_t main_thread_event_id = 0;
-    onnxruntime::contrib::OrtEventPool::GetInstance().SignalEvent(main_thread_event_id);
+    if (onnxruntime::contrib::OrtTasks::GetInstance().ForwardOutputsIsValid()) {
+      ORT_ENFORCE(!status.IsOK());
+      onnxruntime::contrib::OrtTasks::GetInstance().SetForwardOutputs(status, {});
+    }
   },
-                                 std::move(task_.bg_thread_promise_));
+                               std::move(setup_future));
+
+  run_id = std::hash<std::thread::id>()(bg_thread.get_id());
+  bg_threads_[run_id] = std::move(bg_thread);
+  onnxruntime::contrib::OrtTasks::GetInstance().CreateBackgroundTask(run_id);
```

Comment:
Is my understanding that each forward call will create a new thread (and along with it a ```run_id```)?

Reply:
yes.

Reply:
Is it possible then that the ```bg_threads_``` has a lot of entries that are no longer relevant (and hence we need to purge the map once the thread has been joined)?